### ISSUE TITLE
Wrong generated SQL for SQL Server 2012/2014 (DISTINCT / OFFSET)

### ISSUE THREAD

**Issue by hardhub** [linq2db/linq2db#437]
I am using this provider:

SqlServerDataProvider(ProviderName.SqlServer2012, SqlServerVersion.v2012);

Which produces the following SQL query (simplified for demo):

``` SQL
exec sp_executesql N'SELECT DISTINCT
    [test].[Id]
FROM
    [dbo].[Test] [test]
ORDER BY
    [test].[Time] DESC,
    [test].[Id] DESC
OFFSET 0 ROWS FETCH NEXT 60 ROWS ONLY'
```

But it is not valid SQL:
"ORDER BY items must appear in the select list if SELECT DISTINCT is specified."

Linq:

``` C#
var query = from test in storage.Tests
            orderby test.Time descending, test.Id descending
            select test.Id;

var tests = query.Distinct().Page(page, size).ToList();

```

Internally Page method uses Skip and Take to get desired page.

When I am using 2008 provider, it generates ROW_NUMBER and all works fine:

``` SQL
exec sp_executesql N'SELECT *
FROM
(
    SELECT
        t.*,
        ROW_NUMBER() OVER
        (
            ORDER BY
                oby DESC,
                oby1 DESC
        ) as rn
    FROM
    (
        SELECT DISTINCT
            [test].[Id],
            [test].[Time] as [oby],
            [test].[Id] as [oby1]
        FROM
            [dbo].[Test] [test]
    ) t
) t1
WHERE
    t1.rn BETWEEN 1 AND 60'

```

In this specific case (when Id of table is used) I can implement workaround by including Time in query and by wrapping offset query with Id selection. But it is only possible because (Id, Time) will be always unique due to Id.
But in common scenario there are no way to order rows and select page if ordering columns are not targets in query.

So it is not good that something can work for SQL Server 2008 and cannot work for 2012/2014 one.

**Comment by sdanyliv** [comment]
Usually Distinct is another sorting, so your ordering will be ignored.
Looks like for this specific scenario linq2db should also use Window Functions.

**Comment by maxim-boligatov** [comment]
> Usually Distinct is another sorting, so your ordering will be ignored.
> Looks like for this specific scenario linq2db should also use Window Functions.

Is it the comment about better way to implement it?

I think window function also can be used as "fallback" for distinct quires without paging:

```sql
SELECT
    t.*,
    ROW_NUMBER() OVER
    (
        ORDER BY
            oby DESC,
            oby1 DESC
    ) as rn
FROM
(
    SELECT DISTINCT
        [test].[Id],
        [test].[Time] as [oby],
        [test].[Id] as [oby1]
    FROM
        [dbo].[Test] [test]
) t

```
In case of existing index for Time it will have elegant execution plan:

![Actual Execution Plan](https://cloud.githubusercontent.com/assets/4306334/19241022/7dd5f746-8f15-11e6-907a-5a8bedef596a.png)

Fallback should be applied if order by clause (where expression) has columns not included in target selection (new expression).

**Comment by sdanyliv** [comment]
We should implement window functions support #542 before implementing this request.

**Comment by ili** [comment]
@sdanyliv why we do need window function?

we can use subquery for ex:

```sql
SELECT DISTINCT
    [t].[Id],
FROM
(
    SELECT DISTINCT
        [test].[Id],
        [test].[Time]
    FROM
        [dbo].[Test] [test]
    ORDER BY
        [test].[Time] DESC,
        [test].[Id] DESC
) [t]
ORDER BY
    [t].[Id] DESC
OFFSET 0 ROWS FETCH NEXT 60 ROWS ONLY
```

**Comment by sdanyliv** [comment]
@ili, it is not equivalent queries. You've missed proper ordering by Time in result query.

**Comment by ili** [comment]
D'oh!

**Comment by sdanyliv** [comment]
I think we can make quick fix in SQLBuilder that detects Distinct and applies previous paging implementation. It is workaround but without Window Functions support in AST it is only one way.

**Comment by hardhub** [comment]
> I think we can make quick fix in SQLBuilder that detects Distinct and applies previous paging implementation. It is workaround but without Window Functions support in AST it is only one way.

Previously I wrote the following:
> Fallback should be applied if "order by" clause ("where" expression) has columns not included in target selection ("new" expression).

"Distinct and offset" will work if column is in distinct list. But if changing it back to previous version for **all distinct** queries is faster to implement then I agree.

**Comment by ili** [comment]
@sdanyliv i don't think that SqlBuilder is a place to fix... SqlBuilser should build SQL query text from SQL AST, it should not take care of AST...

**Comment by sdanyliv** [comment]
But currently it does. AST is not ready, but we can create temporary workaround using previous implementation as in BasicSQLBuilder.

**Comment by hardhub** [comment]
It was implemented in v1.9.0 with bugs:

``` C#
var query = from test in storage.Tests
            join ...
            join ...
            join ...
            where ...
            where ...
            where ...
            orderby test.Time descending, test.Id descending
            select test.Id;

var tests = query.Distinct().Page(page, size).ToList(); // Page is Skip/Take combination.
```
Sorting is not considered in SQL:
```sql
SELECT
		t.*,
		ROW_NUMBER() OVER
		(
			ORDER BY
				t.[Id1]
		) as rn
	FROM
```

**Comment by sdanyliv** [comment]
@hardhub, i'll try to handle such situations but, in my opinion it's not correct linq query.
Such query will also fail with linq to objects - **Distinct destroys sorting**.
There are a lot of use cases that developer can express by linq but it not easily or never fits to SQL.

Can you try such query? I've used our tables but it easily translates to your sample.
```cs
var page = 0;
var pageSize = 100;

var inner = from p in db.Parent
	select new
	{
		p.ParentID,
		RowNumber = Sql.Ext.RowNumber().Over().OrderByDesc(p.Value1)
			.ThenByDesc(p.ParentID).ToValue(),
	};

var query = inner
	.Distinct()
	.OrderBy(p => p.RowNumber)
	.Select(p => p.ParentID)
	.Skip(page * pageSize).Take(pageSize);
```
It produces the following query, which is more correct in your situation:
```sql
SELECT
	[p].[ParentID] as [ParentID1]
FROM
	(
		SELECT DISTINCT
			[t1].[ParentID],
			ROW_NUMBER() OVER(ORDER BY [t1].[Value1] DESC, [t1].[ParentID] DESC) as [c1]
		FROM
			[Parent] [t1]
	) [p]
ORDER BY
	[p].[c1]
OFFSET 0 ROWS FETCH NEXT 100 ROWS ONLY 
```

**Comment by hardhub** [comment]
>Distinct destroys sorting.

Yes I agree. But not for windows functions!
You can try Linq2db 1.7 version.. It will work if you will specify to use 2008 provider.

I have changed my code to use group by syntax.   
So it is not just related to my case.  
But it looks like some sort of regression for library. 

It is big issue for Linq2db during each next upgrade you never know will it work or not.
And this is required a lot of database-dependent unit tests (which too much expensive for us).
I think it could raise exception in all cases which cannot be correctly translated.
In other words there are no some strict rules what can be do and what not.

**Comment by sdanyliv** [comment]
I meant, that It is accidentally that your linq query works with previous version of DataProvider. Usually we are trying to be closer to linq to object behavior.

**Comment by hardhub** [comment]
@sdanyliv 

>It is accidentally that your linq query works with previous version of DataProvider

I am not sure that you understand all exactly.   
OFFSET/LIMIT is not available for SQL Server 2008 R2.   
So Linq2db uses ROW_NUMBER to get paging on server side.   
Version 1.9 also requires 2008 provider to use ROW_NUMBER instead OFFSET/LIMIT.   
So we just have to use 2008 provider to be compatible with our RDBMS.   

In other words v1.7 and v1.9 with SQL Server 2008 R2 will use ROW_NUMBER.   
But v1.9 ignores sorting in OVER ORDER BY clause and v1.7 does not.

As said above it is better to raise exception if it cannot translate linq-query to sql-query exactly (instead of fetching unexpected data). I personally think it makes sense to throw on each wrong linq-query. It will be much more predictable... Just my thoughts.

**Comment by sdanyliv** [comment]
I have thoughts about that. It is good to have warning or exception i that case? It is ok for you if we introduce setting, for example ControlOrdering. If linq query do not support exact ordering that defined in linq query, we can throw exception. Not sure that we can handle all use cases, but you can help us to find most of them.

**Comment by hardhub** [comment]
> It is good to have warning or exception i that case?
If it is a question... then I think so.  
At least it is principle which we are using.  
As soon as software knows that something is wrong and that cannot be solved with the same success level (as it is in normal case) then the current operation should be halted immediately (like kernel panic). So exception will be a result of operation and it is exactly linq2db responsibility. And I am not not only about ordering linq-queries... I am about all queries.

3 type of checking correctness of code:
1) We can check all on compile time.
2) We can check all during run time.
3) We can do nothing and hope that another logic/tester will find a problem.

1st is impossible (or at least not easy), 3rd is what we have now, 2nd is what we can have with minimal effort.

**Comment by igor-tkachev** [comment]
Check out this example without ```OFFSET 0 ROWS FETCH NEXT 60 ROWS ONLY```. I guess you will get the same result. So, the problem id DISTINCT/ORDER BY.

**Comment by hardhub** [comment]
@igor-tkachev 

Not sure I understand what are you talking about.
As I wrote above 2008 provider of version 1.7 works fine using ROW_NUMBER, but not version 1.9.

**Comment by MaceWindu** [comment]
Fixed by #1221
